From ac9251e4cd4aec4fb62cfe973637ee5e3ec7a347 Mon Sep 17 00:00:00 2001
From: Fangrui Song <maskray@google.com>
Date: Thu, 26 Sep 2019 17:10:09 +0000
Subject: [UPSTREAM][ELF] Set SectionBase::partition in processSectionCommands

Fixes PR43461 (regression caused by D67504)

The partition field of a SECTIONS-specified section is not set after
D67504. The 0 value affects findSection() which checks if the partition
field is 1.

So `Out::initArray = findSection(".init_array")` is null, and
DT_INIT_ARRAYSZ is not set.

Reviewed By: peter.smith

Differential Revision: https://reviews.llvm.org/D68087

llvm-svn: 372996
(cherry picked from commit f1e1451946502dd8f808c251bf4c3b571e80fd9d)

Change-Id: Ibdf629ea56f9200dfcc08c532e27b06b2cfe5abc
---
 lld/ELF/LinkerScript.cpp            |  5 +++++
 lld/test/ELF/linkerscript/dynamic.s | 10 +++++++---
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/lld/ELF/LinkerScript.cpp b/lld/ELF/LinkerScript.cpp
index 74fae363fe4..533f2f6ad73 100644
--- a/lld/ELF/LinkerScript.cpp
+++ b/lld/ELF/LinkerScript.cpp
@@ -513,6 +513,11 @@ void LinkerScript::processSectionCommands() {
           s->alignment = subalign;
       }
 
+      // Set the partition field the same way OutputSection::recordSection()
+      // does. Partitions cannot be used with the SECTIONS command, so this is
+      // always 1.
+      sec->partition = 1;
+
       sec->sectionIndex = i++;
     }
   }
diff --git a/lld/test/ELF/linkerscript/dynamic.s b/lld/test/ELF/linkerscript/dynamic.s
index d236fba77b0..c69e752f9ad 100644
--- a/lld/test/ELF/linkerscript/dynamic.s
+++ b/lld/test/ELF/linkerscript/dynamic.s
@@ -1,10 +1,14 @@
 # REQUIRES: x86
+
+## Test that DT_INIT_ARRAYSZ/DT_FINI_ARRAYSZ/DT_PREINIT_ARRAYSZ are computed
+## correctly, no matter their associated sections are orphans or not.
+
 # RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux %s -o %t1.o
 # RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux %p/Inputs/shared.s -o %t2.o
-# RUN: ld.lld -shared %t2.o -o %t2.so
+# RUN: ld.lld -shared %t2.o -soname=so -o %t2.so
 
-# RUN: echo "SECTIONS { }" > %t.script
-# RUN: ld.lld %t1.o %t2.so -o %t
+# RUN: echo "SECTIONS { .init_array : { *(.init_array) } }" > %t.script
+# RUN: ld.lld -T %t.script %t1.o %t2.so -o %t
 # RUN: llvm-readobj --dynamic-table %t | FileCheck %s
 
 # CHECK:      DynamicSection [
-- 
2.25.0.341.g760bfbb309-goog

